yarn-project/sequencer-client/src/block_builder/circuit_block_builder.ts

```diff
@@ -462,53 +469,165 @@ export class CircuitBlockBuilder implements BlockBuilder {
+   *    This is problematic, as we cannot use either 0 or 2 as our inclusion proof.
+   *    Why cant we?
+   *      - Index 0 has a val 0 and nextVal of 2. This is NOT enough to prove non inclusion of 2.
+   *      - Our existing tree is in a state where we cannot prove non inclusion of 3.
+   *    We do not provide a non inclusion proof to out circuit, but prompt it to look within the insertion subtree.
+   * 2. Update pending insertion subtree
+   * 3. Insert 3 into pending subtree
+   *
+   * (no inclusion proof provided)
+   *  index     0       2       3       4         5       6       -       -
+   *  -------------------------------------      ----------------------------
+   *  val       0       5      10      15         2       3       -       -
+   *  nextIdx   5       2       3       0         6       2       -       -
+   *  nextVal   2      10      15       0         3       5       -       -
+   *
+   * Inserting 20: (happy path)
+   * 1. Find the low nullifier (15) - provide inculsion proof
+   * 2. Update its pointers
+   * 3. Insert 20 into the pending subtree
+   *
+   *  index     0       2       3       4         5       6       7       -
+   *  -------------------------------------      ----------------------------
+   *  val       0       5      10      15         2       3      20       -
+   *  nextIdx   5       2       3       7         6       2       0       -
+   *  nextVal   2      10      15      20         3       5       0       -
+   *
+   * Inserting 19:
+   * 1. In this case we can find a low nullifier, but we are updating a low nullifier that has already been updated
+   *    We can provide an inclusion proof of this intermediate tree state.
+   * 2. Update its pointers
+   * 3. Insert 19 into the pending subtree
+   *
+   *  index     0       2       3       4         5       6       7       8
+   *  -------------------------------------      ----------------------------
+   *  val       0       5      10      15         2       3      20       19
+   *  nextIdx   5       2       3       8         6       2       0       7
+   *  nextVal   2      10      15      19         3       5       0       20
+   *
+   * Perform subtree insertion
+   *
+   *  index     0       2       3       4       5       6       7       8
+   *  ---------------------------------------------------------------------
+   *  val       0       5      10      15       2       3      20       19
+   *  nextIdx   5       2       3       8       6       2       0       7
+   *  nextVal   2      10      15      19       3       5       0       20
+   *
    * TODO: this implementation will change once the zero value is changed from h(0,0,0). Changes incoming over the next sprint
    * @param leaves Values to insert into the tree
    * @returns
    */
   public async performBaseRollupBatchInsertionProofs(leaves: Buffer[]): Promise<LowNullifierWitnessData[] | undefined> {
```

Comment:
I feel like we should break this function out into a seperate file, inject the db instance and create some test cases for it.

Reply:
+1, but I think it's part of a bigger refactor on the block builder. I can take it once this is merged.